### PR TITLE
sys/stdio_uart: Add warning about use in ISR

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Add a warning that RIOT does not officially support STDIO over UART from interrupt context. Some reasons are given below.

### Testing procedure

Read the documentation.

### CODE REVIEW

**Comment by maribu** [comment]
I think that both @benpicco and @gschorcht have scenarios in which not officially supporting STDIO from ISR would be highly beneficial. But so far, STDIO has been happily used from ISR; mostly however for debugging.

**Comment by kaspar030** [comment]
I'm not sure I agree on this.

To me, even simple outputs from UART are an extremely often used tool.
IMO debugging alone makes stdio from ISR an important use case.

```An UART implementation might choose to not guard an atomic sequence of
    interactions with the hardware by disabling IRQs but using a mutex, e.g.
    because the sequence takes long. Locking a mutex from interrupt
    context will result in a deadlock, if the mutex is already locked.```

If some implementations do this, can they be changed?

```If an ISR is triggered from a power saving mode, some peripherals or
    clock domains might still be offline during that ISR; including the UART
    peripheral. This is a valid implementation choice to allow time critical
    low power scenarios being covered by RIOT.```

They might still be offline *after* the interrupt that caused a wakeup.

**Comment by gschorcht** [comment]
> IMO debugging alone makes stdio from ISR an important use case.

Hm, I had to learn that the `stdio` doesn't work at all in ISRs if the newlib uses real locking functions instead of dummy locking functions.

**Comment by kaspar030** [comment]
> Hm, I had to learn that the `stdio` doesn't work at all in ISRs if the newlib uses real locking functions instead of dummy locking functions.

Yeah. But that's newlib(stdio).

**Comment by maribu** [comment]
> IMO debugging alone makes stdio from ISR an important use case.

The proposal does not forbid users from using STDIO in ISRs. It just points out that people cannot expect this to work every time on every board. Indeed, debugging is especially pointed out as an example of when using STDIO over UART is not an anti-pattern to avoid.

Let's compare the situation to LTO, which is also not officially supported: It is known that LTO can be perfectly reliable for many use cases, but currently it sadly will break some stuff. But people with the right understanding can still use LTO happily. And hopefully the remaining bugs exposed by LTO get eventually fixed so that eventually LTO can be officially supported.

Note that the alternative to not officially supporting STDIO over UART from ISR is to *do officially support* it. And there are systems on which this is just not feasible. 

> They might still be offline _after_ the interrupt that caused a wakeup. IMO this is the case on most platforms, the PLLs need some time to get to the configured clock, which often messes with the first (few) characters after waking up from deep sleep.

This honestly concerns me more that not using STDIO from ISR in production systems. Won't other clock domains be equally affected to this? If e.g. an SPI transfer is occasionally unreliable after a wake up e.g. because a transceiver signaled reception of a frame, a correctly received frame might get corrupted during transfer from the transceiver. This is pretty worrying :-/

**Comment by kaspar030** [comment]
> Let's compare the situation to LTO, which is also not officially supported

Do we accept code that does not work with LTO? I don't think so, we consider that a bug.
So, LTO is kinda "supported", there are just corner cases where it breaks stuff. (are we aware of any that haven't been worked around?).
Also, LTO is an optimization, which puts it into a different class of features. Applications usually work with or without LTO, the latter being somewhat faster or smaller. Only in corner cases (non-LTO doesn't fit or non-LTO is too slow) non-working LTO poses an actual problem.

> Note that the alternative to not officially supporting STDIO over UART from ISR is to _do officially support_ it. And there are systems on which this is just not feasible.

Why is it not feasible?

> Won't other clock domains be equally affected to this? If e.g. an SPI transfer is occasionally unreliable after a wake up

Depends. SPI and I2C are clocked by one side, so if that clock is not as fast as it should be, it doesn't really matter in practice. UART is special in that sense.
HF timers might also pose a problem shortly after wakeup.

**Comment by maribu** [comment]
> (are we aware of any that haven't been worked around?).

As a good RIOT citizen I opened PRs fixing every one I was aware of ;-) But I never tested explicitly with the goal of finding bugs with LTO. I just use it from time to time and fix bugs that I ran into by accident.

It would certainly be nice to drop the experimental flag. Maybe an email in the devel list asking people to file bug reports for every issue they are aware of would be good first step in that direction? (I think people are less likely to file bug reports when there is an experimental flag, as they might assume the developers are aware of the issues.)

> Also, LTO is an optimization, which puts it into a different class of features.

I don't think it is relevant to whether claim full and up to standard support (or "officially support it") for some feature if the feature is important or not. This should be based on the quality solely. E.g. if we would have a known-to-be-broken TLS implementation, no one would suggest to drop the experimental flag just because TLS is important.

(Note: Experimental implies that something is "on the road" to full support. If full support is however not planned or not feasible, "experimental" would be misleading. So I'm against claiming that STDIO form ISR is experimental, as I don't see that universal support will ever land in RIOT.)

> Why is it not feasible?

Because the technical effort required to get it working correctly for every platform, for every valid configuration, and for every valid API use is just not worth it.

E.g. let's say I'm using 9600 baud for UART an a 10 ms Watchdog Timer. If I now print 100 characters to STDOUT over UART in my ISR, the ISR blocks every other code from running for at least ~10.4 ms - thus, resulting in a reset. There is not technically convincing way to work around this. (E.g. implicitly feeding the watchdog during every access to stdio could delay or even prevent the watchdog from triggering when it certainly should.)

Blocking the system for even one millisecond (or 10 chars on a 9600 baud UART) would break assumptions of many drivers. I bet many network drivers wouldn't function properly with their ISR not being executed for 1 ms or more. (I would expect they only drop/corrupt a single frame for every puts/printf from ISR. But if the application logic results in puts/printf from ISR and frames going in/out being synchronized, this still could result in consistent failure.)

So: I think it is safe to say that STDIO from ISR in general is just not feasible. (This does not mean that there are use cases and specific setups in which it works reliable.)

I think we either should document the preconditions of using stdio from ISR in detail, or we just add some general "at your own risk" warning. I'm certainly for the latter, as I find more and more gotchas the longer I think about this.

**Comment by benpicco** [comment]
I think adding a hint in the way of 

> Don't access peripherals inside ISR context unless you know what you are doing

would be good. That's consistent with SPI and I2C.
Even better if it has details of which breakage to expect with what, if that stays up to date.

It would sure be possible to always re-initialize the PLL when when a wake-from sleep interrupt is triggered (which would delay that would delay that interrupt).

But I'm not so sure if anyone will implement it anytime soon, as this fixes a rather niece problem that doesn't have an application yet. So this change is merely documenting the status quo.

Maybe we shouldn't suggest that it must always stays this way :wink:

**Comment by maribu** [comment]
> It would sure be possible to always re-initialize the PLL when when a interrupt has triggered a wake-from sleep, we are still in control of the IRQ handler (which would of course delay that interrupt - depends on the application if that is acceptable).

Note that RIOT is intended to be used as general purpose OS. Doing so would conflict with this.

Btw: What are the use cases for STDIO access during ISR except for debugging? IMO, we should just add the warning and spend our time on more pressing stuff.

**Comment by kaspar030** [comment]
I'm fine with a warning to not use it, or only for debugging. I don't think we should allow this to be not supported (as in, we should treat non-working stdio from ISR a bug).

*if* we had proper gdb support on every board, maybe. but as it is, stdio is an important debugging tool.

I suggest removing the "not officially supported" wording and just put a warning that this should only be used for debugging because its synchronous nature will completely bork all timing assumptions in production code.

File: sys/include/stdio_uart.h [hunk 1]
```diff
@@ -19,6 +19,27 @@
...
+ * event could be posted to an @ref sys_event and the actual STDIO operation
+ * being deferred to thread context.
+ *
+ * Some reasons why RIOT does not officially support STDIO over UART from ISR:
+ * 1. UART is *slow* and the system easily remains in interrupt context for
+ *    unacceptable long time.
+ * 2. An UART implementation might choose to not guard an atomic sequence of
+ *    interactions with the hardware by disabling IRQs but using a mutex, e.g.
+ *    because the sequence takes long. Locking a mutex from interrupt
+ *    context will result in a deadlock, if the mutex is already locked.
```
Comment (maribu) [inline sys/include/stdio_uart.h:37]: This argument becomes invalid by supporting stdio from UART; an implementation must not use a mutex for synchronization in order to allow for stdio from UART for debugging. I will drop it.

Comment (maribu) [reply sys/include/stdio_uart.h:37]: Hmm, a list of two arguments doesn't look good. Now I have to come up with one more good argument :-D

File: sys/include/stdio_uart.h [hunk 2]
```diff
@@ -19,6 +19,34 @@
...
+ * enough and for some platforms even reliable. Production code however should
+ * fully avoid any access to STDIO from interrupt context. Instead, e.g. an
+ * event could be posted to an @ref sys_event and the actual STDIO operation
+ * being deferred to thread context.
+ *
+ * Some reasons why STDIO over UART from ISR should be avoided:
+ * 1. UART is *slow* and the system easily remains in interrupt context for
+ *    unacceptable long time.
+ *     - E.g. sending 100 chars at 9600 baud will block the system for
+ *       10 milli seconds.
```
Comment (kaspar030) [inline sys/include/stdio_uart.h:35]:
```suggestion
 *       100 milli seconds.
```

9600 baud are ~1000b/sec...

Comment (kaspar030) [reply sys/include/stdio_uart.h:35]: maybe remove the space also, millisecond is a word ;)

File: sys/include/stdio_uart.h [hunk 3]
```diff
@@ -19,6 +19,34 @@
...
+ *     - Missed deadlines, lost interrupts, or watch dog timer resets can easily
+ *       be caused by this.
+ * 2. Even if DMA is used for UART, using STDIO form ISR can cause significant
```
Comment (kaspar030) [inline sys/include/stdio_uart.h:38]:
```suggestion
 * 2. Even if DMA is used for UART, using STDIO from ISR can cause significant
```

Comment (kaspar030) [reply sys/include/stdio_uart.h:38]: or mabe "within"

**Comment by kaspar030** [comment]
sorry for chipping in, found some typos. feel free to ignore, definitely squash directly.

File: sys/include/stdio_uart.h (hunk 2 above)
Comment (maribu) [reply sys/include/stdio_uart.h:35]: 9,600 is closer to 10,000 than to 1,000 ;-)

100 / 9600 baud = 0.01042s ~= 10ms

Comment (kaspar030) [reply sys/include/stdio_uart.h]: yeah, it is closer to 10000, but a baud with only two symbols (0 and 1) is closer to bit than to char... (100 * 8) / 9600 = 0.0833333 ~= 83ms.

Comment (maribu) [reply sys/include/stdio_uart.h]: I always assumed the whole 8 data bits + parity and stop bits would count as a symbol :-/

Fixed. Thanks!

**Comment by maribu** [comment]
Thanks for the reviews!

**Pull request merged**